File: Source/GraphAlgorithms/is_tree.js
let visited = [];

function DFS(v, parent, adjList) {
    visited[v] = true;
    for(let i = 0; i < adjList[v].length; i++) {
        if(!visited[adjList[v][i]]) {
            if(DFS(adjList[v][i], v, adjList)) {
                return true;
            }
        } else if(adjList[v][i] != parent) {
            return true;
        }
    }
    return false;
}

function isTree(adjList) {
    visited = new Array(adjList.length).fill(false);
    if(DFS(0, -1, adjList)) {
        return false;
    }
    for(let i = 0; i < adjList.length; i++) {
        if(!visited[i]) {
            return false;
        }
    }
    return true;
}

File: Source/GraphAlgorithms/is_tree.c
#include <stdio.h>

#define MAX 100

int visited[MAX];
int matrix[MAX][MAX];

int DFS(int v, int parent, int n) {
    visited[v] = 1;
    for(int i = 0; i < n; i++) {
        if(matrix[v][i] == 1) {
            if(!visited[i]) {
                if(DFS(i, v, n)) {
                    return 1;
                }
            } else if(i != parent) {
                return 1;
            }
        }
    }
    return 0;
}

int isTree(int n) {
    if(DFS(0, -1, n)) {
        return 0;
    }
    for(int i = 0; i < n; i++) {
        if(!visited[i]) {
            return 0;
        }
    }
    return 1;
}
